application/migrations/231_adif_315.php

```diff
@@ -8,542 +8,574 @@
+		if (!$col_check) {
+			$qso_fields[] = "ALTER TABLE `$table_name` ADD COLUMN COL_DCL_QSLSDATE DATETIME DEFAULT NULL AFTER COL_DCL_QSL_RCVD;";
+		} else {
+			log_message('info', 'Column "COL_DCL_QSLSDATE" already exists, skipping ALTER TABLE.');
+		}
+
+		// Add DCL_QSL_SENT
+		// https://adif.org/315/ADIF_315_annotated.htm#QSO_Field_DCL_QSL_SENT
+		$col_check = $this->db->query("SHOW COLUMNS FROM `$table_name` LIKE 'COL_DCL_QSL_SENT';")->num_rows() > 0;
+		if (!$col_check) {
+			$qso_fields[] = "ALTER TABLE `$table_name` ADD COLUMN COL_DCL_QSL_SENT VARCHAR(10) DEFAULT NULL AFTER COL_DCL_QSLSDATE;";
+		} else {
+			log_message('info', 'Column "COL_DCL_QSL_SENT" already exists, skipping ALTER TABLE.');
+		}
+
+		// Add MORSE_KEY_INFO
+		// https://adif.org/315/ADIF_315_annotated.htm#QSO_Field_MORSE_KEY_INFO
+		$col_check = $this->db->query("SHOW COLUMNS FROM `$table_name` LIKE 'COL_MORSE_KEY_INFO';")->num_rows() > 0;
+		if (!$col_check) {
+			$qso_fields[] = "ALTER TABLE `$table_name` ADD COLUMN COL_MORSE_KEY_INFO VARCHAR(255) DEFAULT NULL AFTER COL_REGION;";
+		} else {
+			log_message('info', 'Column "COL_MORSE_KEY_INFO" already exists, skipping ALTER TABLE.');
+		}
+
+		// Add MORSE_KEY_TYPE
+		// https://adif.org/315/ADIF_315_annotated.htm#QSO_Field_MORSE_KEY_TYPE
+		$col_check = $this->db->query("SHOW COLUMNS FROM `$table_name` LIKE 'COL_MORSE_KEY_TYPE';")->num_rows() > 0;
+		if (!$col_check) {
+			$qso_fields[] = "ALTER TABLE `$table_name` ADD COLUMN COL_MORSE_KEY_TYPE VARCHAR(10) DEFAULT NULL AFTER COL_MORSE_KEY_INFO;";
+		} else {
+			log_message('info', 'Column "COL_MORSE_KEY_TYPE" already exists, skipping ALTER TABLE.');
+		}
+
+		// Add QSLMSG_RCVD
+		// https://adif.org/315/ADIF_315_annotated.htm#QSO_Field_QSLMSG_RCVD
+		$col_check = $this->db->query("SHOW COLUMNS FROM `$table_name` LIKE 'COL_QSLMSG_RCVD';")->num_rows() > 0;
+		if (!$col_check) {
+			$qso_fields[] = "ALTER TABLE `$table_name` ADD COLUMN COL_QSLMSG_RCVD VARCHAR(255) DEFAULT NULL AFTER COL_QSLMSG;";
+		} else {
+			log_message('info', 'Column "COL_QSLMSG_RCVD" already exists, skipping ALTER TABLE.');
+		}
+
+		// Run the querys
+		try {
+			foreach ($qso_fields as $query) {
+				$this->db->query($query);
+			}
+		} catch (Exception $e) {
+			$this->db->trans_rollback();
+			log_message('error', 'Migration failed: ' . $e->getMessage());
+			log_message('error', 'The query was: ' . $query);
+			return false;
+		}
+
+		// ############################################################################################################
+
+		/**
+		 * 2. Primary Subdivisions
+		 * 
+		 * https://adif.org/315/ADIF_315_annotated.htm#Primary_Administrative_Subdivision
+		 */
+
+		// Clean Up primary_subdivisions from dupes, since there's no PK!!
+		$this->db->query("CREATE TABLE tmp_psd AS SELECT DISTINCT ps.`adif`,ps.`state`,ps.`subdivision`,ps.`deprecated` FROM `primary_subdivisions` ps");
```

Comment:
At this point in Code the column `ps.deprecated` does not exist. It will be created in Line 156